File: src/main/java/fi/haapatalo/android/eyeball/EyeballViewerActivity.java
/*
 * Copyright 2017 Google Inc. All Rights Reserved.

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fi.haapatalo.android.eyeball;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.widget.Toast;
import com.google.vr.sdk.controller.Controller;
import com.google.vr.sdk.controller.ControllerManager;
import com.google.vr.sdk.widgets.pano.VrPanoramaEventListener;
import com.google.vr.sdk.widgets.pano.VrPanoramaView;
import com.google.vr.sdk.widgets.pano.VrPanoramaView.Options;

import java.io.File;

/**
 * A basic PanoWidget Activity to load panorama images from disk. It will load a test image by
 * default. It can also load an arbitrary image from disk using:
 * adb shell am start -a "android.intent.action.VIEW" \
 * -n "com.google.vr.sdk.samples.simplepanowidget/.EyeballViewerActivity" \
 * -d "/sdcard/FILENAME.JPG"
 * <p>
 * To load stereo images, "--ei inputType 2" can be used to pass in an integer extra which will set
 * VrPanoramaView.Options.inputType.
 */
public class EyeballViewerActivity extends Activity {
    private static final String TAG = EyeballViewerActivity.class.getSimpleName();
    /**
     * Actual panorama widget.
     **/
    private VrPanoramaView panoWidgetView;
    /**
     * Arbitrary variable to track load status. In this example, this variable should only be accessed
     * on the UI thread. In a real app, this variable would be code that performs some UI actions when
     * the panorama is fully loaded.
     */
    public boolean loadImageSuccessful;
    private ImageLoaderTask backgroundImageLoaderTask;

    // These two objects are the primary APIs for interacting with the Daydream controller.
    private ControllerManager controllerManager;
    private Controller controller;

    // The various events we need to handle happen on arbitrary threads. They need to be reposted to
    // the UI thread in order to manipulate the TextViews. This is only required if your app needs to
    // perform actions on the UI thread in response to controller events.
    private Handler uiHandler = new Handler();

    private ImageBrowser browser;

    /**
     * Called when the app is launched via the app icon or an intent using the adb command above. This
     * initializes the app and loads the image to render.
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.main_layout);

        panoWidgetView = (VrPanoramaView) findViewById(R.id.pano_view);
        panoWidgetView.setEventListener(new ActivityEventListener());

        // Start the ControllerManager and acquire a Controller object which represents a single
        // physical controller. Bind our listener to the ControllerManager and Controller.
        EventListener listener = new EventListener();
        controllerManager = new ControllerManager(this, listener);
        controller = controllerManager.getController();
        controller.setEventListener(listener);

        browser = new ImageBrowser(new ImageBrowser.ImageLoader() {
            @Override
            public void load(File image) {
                loadImage(image);
            }
        });
        // Initial launch of the app or an Activity recreation due to rotation.
        handleIntent(getIntent());
    }

    /**
     * Called when the Activity is already running and it's given a new intent.
     */
    @Override
    protected void onNewIntent(Intent intent) {
        Log.i(TAG, this.hashCode() + ".onNewIntent()");
        // Save the intent. This allows the getIntent() call in onCreate() to use this new Intent during
        // future invocations.
        setIntent(intent);
        // Load the new image.
        handleIntent(intent);
    }

    /**
     * Load custom images based on the Intent or load the default image. See the Javadoc for this
     * class for information on generating a custom intent via adb.
     */
    private void handleIntent(Intent intent) {
        // Load the bitmap in a background thread to avoid blocking the UI thread. This operation can
        // take 100s of milliseconds.
        browser.load();
    }

    private void loadImage(File image) {
        if (image == null) {
            Log.w(TAG, "No image specified!");
            Toast.makeText(
                    EyeballViewerActivity.this, "No images found in /Pictures/Eyeball", Toast.LENGTH_LONG)
                    .show();
            return;
        }
        Log.i(TAG, "Preparing to load image " + image);
        if (backgroundImageLoaderTask != null) {
            // Cancel any task from a previous intent sent to this activity.
            backgroundImageLoaderTask.cancel(true);
        }
        backgroundImageLoaderTask = new ImageLoaderTask();
        backgroundImageLoaderTask.execute(image);
    }

    @Override
    protected void onStart() {
        super.onStart();
        controllerManager.start();
    }

    @Override
    protected void onStop() {
        controllerManager.stop();
        super.onStop();
    }

    @Override
    protected void onPause() {
        panoWidgetView.pauseRendering();
        super.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        panoWidgetView.resumeRendering();
    }

    @Override
    protected void onDestroy() {
        // Destroy the widget and free memory.
        panoWidgetView.shutdown();

        // The background task has a 5 second timeout so it can potentially stay alive for 5 seconds
        // after the activity is destroyed unless it is explicitly cancelled.
        if (backgroundImageLoaderTask != null) {
            backgroundImageLoaderTask.cancel(true);
        }
        super.onDestroy();
    }

    /**
     * Helper class to manage threading.
     */
    class ImageLoaderTask extends AsyncTask<File, Void, Boolean> {

        /**
         * Reads the bitmap from disk in the background and waits until it's loaded by pano widget.
         */
        @Override
        protected Boolean doInBackground(File... fileInformation) {
            return loadImage(fileInformation[0]);
        }

        protected Boolean loadImage(File image) {
            Log.i(TAG, "Loading image: " + image);
            Options panoOptions = new Options();
            panoOptions.inputType = Options.TYPE_STEREO_OVER_UNDER;
            Bitmap bmp = BitmapFactory.decodeFile(image.getPath());
            panoWidgetView.loadImageFromBitmap(bmp, panoOptions);
            return true;
        }
    }

    /**
     * Listen to the important events from widget.
     */
    private class ActivityEventListener extends VrPanoramaEventListener {
        /**
         * Called by pano widget on the UI thread when it's done loading the image.
         */
        @Override
        public void onLoadSuccess() {
            loadImageSuccessful = true;
        }

        /**
         * Called by pano widget on the UI thread on any asynchronous error.
         */
        @Override
        public void onLoadError(String errorMessage) {
            loadImageSuccessful = false;
            Toast.makeText(
                    EyeballViewerActivity.this, "Error loading pano: " + errorMessage, Toast.LENGTH_LONG)
                    .show();
            Log.e(TAG, "Error loading pano: " + errorMessage);
        }
    }


    // We receive all events from the Controller through this listener. In this example, our
    // listener handles both ControllerManager.EventListener and Controller.EventListener events.
    // This class is also a Runnable since the events will be reposted to the UI thread.
    private class EventListener extends Controller.EventListener
            implements ControllerManager.EventListener, Runnable {

        @Override
        public void onApiStatusChanged(int state) {
            uiHandler.post(this);
        }

        @Override
        public void onConnectionStateChanged(int state) {
            uiHandler.post(this);
        }

        @Override
        public void onRecentered() {
            // In a real GVR application, this would have implicitly called recenterHeadTracker().
            // Most apps don't care about this, but apps that want to implement custom behavior when a
            // recentering occurs should use this callback.
            Log.d(TAG, "Recentered");
        }

        @Override
        public void onUpdate() {
            uiHandler.post(this);
        }

        private boolean clickButtonState = false;
        private boolean appButtonState = false;

        // Update the various TextViews in the UI thread.
        @Override
        public void run() {
            controller.update();

            if (controller.appButtonState != appButtonState) {
                appButtonState = controller.appButtonState;
                if (appButtonState) {
                    Log.i(TAG, "App button pressed, switching to previous image");
                    browser.previous();
                }
            }
            if (controller.clickButtonState != clickButtonState) {
                clickButtonState = controller.clickButtonState;
                if (clickButtonState) {
                    Log.i(TAG, "Click button pressed, switching to next image");
                    browser.next();
                }
            }
        }
    }
}

File: src/main/java/fi/haapatalo/android/eyeball/EyeballViewerActivity.kt
package fi.haapatalo.android.eyeball

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import com.google.vr.sdk.widgets.pano.VrPanoramaView
import com.google.vr.sdk.widgets.video.VrVideoView

class EyeballViewerActivity : Activity() {

    // The various events we need to handle happen on arbitrary threads. They need to be reposted to
    // the UI thread in order to manipulate the TextViews. This is only required if your app needs to
    // perform actions on the UI thread in response to controller events.
    private val uiHandler = Handler()

    // All state is managed in these classes that maintain non-null values
    private var imageViewer: ImageViewer? = null
    private var videoViewer: VideoViewer? = null
    private var controller: ControllerHandler? = null
    private val dirs = ImageBrowser.directories

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_layout)

        val panoView = findViewById(R.id.pano_view) as VrPanoramaView
        val videoView = findViewById(R.id.video_view) as VrVideoView

        imageViewer = ImageViewer(this, uiHandler, panoView)
        videoViewer = VideoViewer(this, uiHandler, videoView)

        val control = object: EyeballController {
            override fun next() {
                imageViewer?.next()
                videoViewer?.next()
            }

            override fun previous() {
                imageViewer?.previous()
                videoViewer?.previous()
            }
        }
        controller = ControllerHandler(this, uiHandler, control)

        val dirSpinner = findViewById(R.id.dir_selector) as Spinner
        val dirAdapter = ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, dirs.map { it.first })
        dirSpinner.adapter = dirAdapter
        dirSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) {
            }

            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                imageViewer?.select(dirs[position].second)
                videoViewer?.select(dirs[position].second)
            }
        }

        // Initial launch of the app or an Activity recreation due to rotation.
        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        Log.i(TAG, "${hashCode()}.onNewIntent()")
        // Save the intent. This allows the getIntent() call in onCreate() to use this new Intent during
        // future invocations.
        setIntent(intent)
        // Load the new image.
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        Log.i(TAG, "Start from intent $intent")
        imageViewer?.load()
        videoViewer?.load()
    }

    override fun onStart() {
        super.onStart()
        controller?.onStart()
    }

    override fun onStop() {
        controller?.onStop()
        super.onStop()
    }

    override fun onPause() {
        imageViewer?.onPause()
        videoViewer?.onPause()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        imageViewer?.onResume()
        videoViewer?.onResume()
    }

    override fun onDestroy() {
        imageViewer?.onDestroy()
        videoViewer?.onDestroy()
        super.onDestroy()
    }

    companion object {
        private val TAG = EyeballViewerActivity::class.java.simpleName
    }
}
